etc/keepalived/kp_local_loop.py

```diff
@@ -0,0 +1,665 @@
+            update_dict['status'] = 'up'
+        zk_cli.update_node(self.name, **update_dict)
+        self.old_heartbeat = hb
+
+    def is_check_heart_beat_overtime(self):
+        try:
+            over_time = parse_isotime(
+                self.old_heartbeat) + datetime.timedelta(
+                seconds=self.hb_interval_time)
+        except ValueError:
+            # The heartbeat is not formatted, this must be the kp on the node
+            # is not finish the first loop. We just return False, as we are in
+            # the initializing process.
+            return False
+        current_time = datetime.datetime.utcnow().replace(tzinfo=iso8601.UTC)
+        return current_time > over_time
+
+    def is_necessary_service(self, service_name):
+        return service_name in self.necessary_service_names
+
+    def is_maintained(self):
+        return self.status == 'maintaining'
+
+    def post_alarmed_if_possible(self):
+        if not self.alarmed:
+            zk_cli = get_zk_cli()
+            zk_cli.update_node(self.name, alarmed=True)
+            self.alarmed = True
+
+
+def get_zk_cli():
+    global GLOBAL_ZK
+    if not GLOBAL_ZK:
+        cfg = configparser.ConfigParser()
+        cfg.read(ZK_CLI_CONF)
+        zk_cli = zk.ZooKeeper(cfg)
+        zk_cli.connect()
+        GLOBAL_ZK = zk_cli
+    return GLOBAL_ZK
+
+
+def load_from_zk():
+    zk_cli = get_zk_cli()
+    node_name = get_local_hostname()
+    zk_cli.connect()
+    zk_node = zk_cli.get_node(node_name)
+    local_node = Node(zk_node['name'], zk_node['role'],
+                      zk_node['type'], zk_node['ip'],
+                      zk_node['heartbeat'], zk_node['alarmed'],
+                      zk_node['status'])
+
+    for service in zk_cli.list_services(node_name_filter=node_name):
+        local_node.services.append(
+            Service(service['name'], service['alarmed'],
+                    service['alarmed_at'], service['restarted'],
+                    service['restarted_at'], service['is_necessary'],
+                    service['node_role'], local_node.type,
+                    service['status']))
+        if service['is_necessary']:
+            local_node.necessary_service_names.append(service['name'])
+
+    global GLOBAL_NODE
+    GLOBAL_NODE = local_node
+    return GLOBAL_NODE
+
+
+def get_local_node():
+    global GLOBAL_NODE
+    if not GLOBAL_NODE:
+        return load_from_zk()
+    return GLOBAL_NODE
+
+
+def process():
+    local_node = get_local_node()
+    script_load_pre_check(local_node)
+    if local_node.status in ['maintaining', 'down']:
+        return
+    local_node_service_process(local_node)
+    if local_node.role != 'zookeeper':
+        oppo_node_process()
+
+
+def local_node_service_process(node_obj):
+    for service_obj in node_obj.services:
+        treat_single_service(service_obj, node_obj)
+
+    node_obj.report_heart_beat()
+
+
+def treat_single_service(service_obj, node_obj):
+    if node_obj.is_maintained():
```

Comment:
this can  be removed now. so as `is_maintained()`

Reply:
done